SCClassLibrary/External/Ableton/SCClockMeterSync.sc

```diff
@@ -0,0 +1,246 @@
+// baseBarBeat_ (e.g. MIDISyncClocks, or BeaconClock)
+
+// TODO OSC message prefix with SC_
+
+SCClockMeterSync {
+	var <clock, <>id, <ports;
+	var <repeats = 4, <delta = 0.01, lastReceived;
+	var addrs, meterChangeResp, meterQueryResp;
+
+	// normally clock.setMeterAtBeat notifies \meter
+	// and this watcher should broadcast the change.
+	// but the watcher may need to change baseBarBeat internally.
+	// this should not broadcast. flag is off in that case
+	var broadcastMeter = true;
+
+	*new { |clock, id, ports|
+		^super.new.init(clock, id, ports)
+	}
+
+	init { |argClock, argId, argPorts|
+		clock = argClock;
+		id = argId ?? { 0x7FFFFFFF.rand };
+		this.ports = argPorts;  // see ports_ below
+		meterQueryResp = OSCFunc({ |msg|
+			// because of redundancy ('repeats'), we may receive this message
+			// multiple times. Respond only once.
+			if(msg != lastReceived) {
+				lastReceived = msg;
+				this.prBroadcast(
+					'/SC_LinkClock/meterReply', id, msg[1], this.enabled.asInteger,
+					clock.beats, clock.beatsPerBar, clock.baseBarBeat, clock.beatInBar
+				);
+			};
+		}, '/SC_LinkClock/queryMeter');
+		clock.addDependant(this);
+		this.enabled = true;  // assume enabled when creating
+	}
+
+	free {
+		clock.removeDependant(this);
+		meterQueryResp.free;
+		meterChangeResp.free;
+	}
+
+	ports_ { |argPorts|
+		ports = (argPorts ?? { (57120 .. 57127) }).asArray;
+		addrs = ports.collect { |port| NetAddr("255.255.255.255", port) };
+	}
+
+	enabled { ^meterChangeResp.notNil }
+
+	enabled_ { |bool|
+		if(bool) {
+			if(clock.numPeers == 0) {
+				// maybe just started the clock; allow a beat for Link to hook up
+				// you would think it should go onto 'this'
+				// but the relationship between beats and seconds is unstable initially
+				// in practice it works much better to schedule the sync on a stable clock
+				SystemClock.sched(1.0, {
+					// clock.numPeers: might have changed in the last 2 seconds
+					this.resyncMeter(verbose: clock.numPeers > 0)
+				});
+			} {
+				this.resyncMeter;
+			};
+			meterChangeResp = OSCFunc({ |msg|
+				if(msg[1] != id and: { msg != lastReceived }) {  // ignore message that I sent
+					lastReceived = msg;
+					// [2] = beatsPerBar, [3] = remote clock's real time, [4] = remote clock's barline
+					// also, 5/8 means maybe setting the barline to a half-beat
+					// but we have to round the barline because OSC receipt time is inexact
+					// -- get the rounding factor from baseBarBeat.asFraction
+					// *and* use my utility method because we do not want to broadcast here
+					this.setMeterAtBeat(msg[2],
+						(clock.beats + msg[4] - msg[3]).round(msg[4].asFraction[1].reciprocal)
+					);
+				};
+			}, '/SC_LinkClock/changeMeter');
+		} {
+			meterChangeResp.free;
+			meterChangeResp = nil;
+		};
+	}
+
+	repeats_ { |value|
+		if(value.isKindOf(SimpleNumber).not or: { value < 1 }) {
+			Error("Invalid number of repeats '%'".format(value)).throw;
+		};
```

Comment:
the value check i understand, but not the class validation.. isn't it more smalltalk-y to allow any object here that works like a number?

Reply:
Ideally yes, but:

```
(
c = Complex(1, 0);

c.isNumber.debug("Complex is a number");

fork { c.wait };
)

Complex is a number: true
-> a Routine
ERROR: Message 'wait' not understood.
RECEIVER:
Instance of Complex
```

We're missing a polymorphic interface for `isWaitDuration`. I'm not quite sure how else to handle it in the context of this PR.

Incidentally SimpleNumber is the only number-like class that defines `wait`.

In the backend, threads are rescheduled if they yield something where `slotDoubleVal()` succeeds. As far as I can see, that eventually gets around to:

```
template <typename numeric_type> inline int slotVal(PyrSlot* slot, numeric_type* value) {
    if (IsFloat(slot)) {
        *value = static_cast<numeric_type>(slot->u.f);
        return errNone;
    } else if (IsInt(slot)) {
        *value = static_cast<numeric_type>(slot->u.i);
        return errNone;
    }
    return errWrongType;
}

... and...
inline bool IsFloat(const PyrSlot* slot) { return slot->tag == tagFloat; }
```

So we should allow only slots that are tagged float or int (i.e. SimpleNumber). Anything else will cause the thread to hang indefinitely. And, `anythingElse.wait` will throw a DoesNotUnderstand error, also causing the thread to hang indefinitely. So, this is ugly, but it might be the safest way to ensure the thread proceeds.

Reply:
that's fair. i guess my concern is, do we want to start auditing every bit of code in the class library for this? when i pointed out that the thread could end up erroring and hanging indefinitely, i wasn't thinking it meant it had to be avoided at all costs. i was noting it to see if you had any thoughts or opinions on it.

if you think this needs to be addressed this way, then i think we need to mitigate technical debt:
1. open a new issue describing the problem
2. agree on a workaround (what you have here seems fine)
3. comment everywhere the workaround is used so that when the issue opened in (1) is resolved, the workarounds can be easily spotted and replaced with the solution
4. in some order - compile a list of places where this safeguard is needed, and solve the underlying issue, and then safeguard unsafe `waits` relying on user-given values

Reply:
I guess, if that's what is required, then we might as well just let it error out. That's more in keeping with SC's general attitude of allowing users to shoot themselves in the feet. (I did misunderstand your comment -- I thought you were asking for prevention.)

It makes me a little curious about garbage collection for hanging threads. In this case, if the wait time is valid, the queryMeter thread is reachable via the clock's queue array. The thread is running an open function whose containing scope has received a callback function from the resyncMeter thread, and that function refers to a Condition which contains the resyncMeter thread. I *suppose*, if the queryMeter thread timeout is not scheduled (by error or by inappropriate yield value), then the whole construct is unreachable and GC-able. If that's the case, then I'm fine with letting it die and then it's up to GC. But if we will have these stray actions hanging about forever, then we should more actively prevent it.

Reply:
> I guess, if that's what is required, then we might as well just let it error out. That's more in keeping with SC's general attitude of allowing users to shoot themselves in the feet. (I did misunderstand your comment -- I thought you were asking for prevention.)

up to you. i think that 'attitude' is less an explicit choice and more often lack of time to make things as robust as we'd like. which is understandable.

> But if we will have these stray actions hanging about forever, then we should more actively prevent it.

they will be reachable since they are not freed, so they will not be garbage collected. you could protect {} the wait and free in the cleanup function.

Reply:
> they will be reachable since they are not freed, so they will not be garbage collected.

Oh gosh, good catch, yes, that should be protect-ed.

My question is really about the threads, though, for which there is no destructor. I would assume that a thread is either executing (can't be GC'ed) or scheduled on a clock (referenced from a queue, can't be GC'ed) or yielded a value that prevents rescheduling, in which case... there might be no references? (Unless the user is keeping one.) I suppose it must be like that; otherwise we'd be leaking memory like crazy.